packages/hardhat-core/test/internal/hardhat-network/provider/TransactionQueue.ts

```diff
@@ -0,0 +1,303 @@
+      it("Should not accept an EIP-1559 tx if no base fee is used", function () {
+        const tx1 = createTestTransaction({
+          maxFeePerGas: 1,
+          maxPriorityFeePerGas: 1,
+        });
+
+        assert.throws(
+          () => new TransactionQueue(makeOrderedTxMap([tx1])),
+          InternalError
+        );
+      });
+
+      describe("Sorting transactions", function () {
+        it("Should use the gasPrice and order to sort txs", function () {
+          const tx1 = createTestTransaction({ gasPrice: 123 });
+
+          const tx2 = createTestTransaction({
+            gasPrice: 1000,
+          });
+
+          // This has the same gasPrice than tx2, but arrived later, so it's
+          // placed later in the queue
+          const tx3 = createTestTransaction({
+            gasPrice: 1000,
+          });
+
+          const tx4 = createTestTransaction({
+            gasPrice: 2000,
+          });
+
+          const queue = new TransactionQueue(
+            makeOrderedTxMap([tx1, tx2, tx3, tx4])
+          );
+
+          assert.equal(queue.getNextTransaction(), tx4.data);
+          assert.equal(queue.getNextTransaction(), tx2.data);
+          assert.equal(queue.getNextTransaction(), tx3.data);
+          assert.equal(queue.getNextTransaction(), tx1.data);
+        });
+      });
+
+      it("Should not include transactions from a sender whose next tx was discarded", function () {
+        const senderWithFirstTxNotMined =
+          "0x0000000000000000000000000000000000000001";
+
+        const senderWithThirdTxNotMined =
+          "0x0000000000000000000000000000000000000002";
+
+        const tx1 = createTestTransaction({
+          gasPrice: 100,
+        });
+
+        const tx2 = createTestTransaction({
+          gasPrice: 99,
+        });
+
+        // Not mined
+        const tx3 = createTestTransaction({
+          gasPrice: 98,
+          from: senderWithFirstTxNotMined,
+        });
+
+        // Discarded
+        const tx4 = createTestTransaction({
+          gasPrice: 97,
+          from: senderWithFirstTxNotMined,
+        });
+
+        const tx5 = createTestTransaction({
+          gasPrice: 96,
+          from: senderWithThirdTxNotMined,
+        });
+
+        const tx6 = createTestTransaction({
+          gasPrice: 95,
+          from: senderWithThirdTxNotMined,
+        });
+
+        // Not mined
+        const tx7 = createTestTransaction({
+          gasPrice: 94,
+          from: senderWithThirdTxNotMined,
+        });
+
+        // Discarded
+        const tx8 = createTestTransaction({
+          gasPrice: 93,
+          from: senderWithThirdTxNotMined,
+        });
+
+        const tx9 = createTestTransaction({
+          gasPrice: 92,
+        });
+
+        const tx10 = createTestTransaction({
+          gasPrice: 91,
+        });
+
+        const queue = new TransactionQueue(
+          makeOrderedTxMap([tx1, tx2, tx3, tx4, tx5, tx6, tx7, tx8, tx9, tx10])
```

Comment:
Something I always think about these kinds of tests... should this array be shuffled?

Reply:
Yes! I'm glad you raised this. I had the exact same thought, and wasn't sure if we should do that.